test/fw/ptl/utils/plugins/ptl_test_data.py

```diff
@@ -55,240 +56,260 @@ class PTLTestData(Plugin):
-        else:
             # test does not have any PBS Objects, so just return
             return
-        if not hasattr(_test, 'server'):
-            # test does not have any PBS Objects, so just return
+        if self.__priv_sn != sn:
+            self.__save_data_count = 0
+            self.__priv_sn = sn
+        # Saving home might take time so disable timeout
+        # handler set by runner
+        tn = getattr(_test, '_testMethodName', 'unknown')
+        testlogs = getattr(test, 'captured_logs', '')
+        datadir = os.path.join(self.post_data_dir, sn, tn)
+        if os.path.exists(datadir):
+            _msg = 'Old post analysis data exists at %s' % datadir
+            _msg += ', skipping saving data for this test case'
+            self.logger.warn(_msg)
+            _msg = 'Please remove old directory or'
+            _msg += ' provide different directory'
+            self.logger.warn(_msg)
             return
-        st = getattr(test, 'start_time', None)
-        if st is not None:
-            st = time.mktime(st.timetuple())
+        if getattr(test, 'old_sigalrm_handler', None) is not None:
+            _h = getattr(test, 'old_sigalrm_handler')
+            signal.signal(signal.SIGALRM, _h)
+            signal.alarm(0)
+        self.logger.log(logging.DEBUG2, 'Saving post analysis data...')
+        current_host = socket.gethostname().split('.')[0]
+        self.du.mkdir(current_host, path=datadir, mode=0755,
+                      parents=True, logerr=False, level=logging.DEBUG2)
+        if err is not None:
+            if isclass(err[0]) and issubclass(err[0], SkipTest):
+                status = 'SKIP'
+                status_data = 'Reason = %s' % (err[1])
+            else:
+                if isclass(err[0]) and issubclass(err[0], TimeOut):
+                    status = 'TIMEDOUT'
+                status_data = getattr(test, 'err_in_string', '')
         else:
-            st = time.time()
-        st -= 180  # starttime - 3 min
-        et = getattr(test, 'end_time', None)
-        if et is not None:
-            et = time.mktime(et.timetuple())
+            status_data = ''
+        logfile = os.path.join(datadir, 'logfile_' + status)
+        f = open(logfile, 'w+')
+        f.write(testlogs + '\n')
+        f.write(status_data + '\n')
+        f.write('test duration: %s\n' % str(getattr(test, 'duration', '0')))
+        if status in ('PASS', 'SKIP'):
+            # Test case passed or skipped, no need to save post analysis data
+            f.close()
+            return
+        if ((self.max_postdata_threshold != 0) and
+                (self.__save_data_count > self.max_postdata_threshold)):
+            _msg = 'Total number of saved post analysis data for this'
+            _msg += ' testsuite is exceeded max postdata threshold'
+            _msg += ' (%d)' % self.max_postdata_threshold
+            f.write(_msg + '\n')
+            self.logger.error(_msg)
+            f.close()
+            return
```

Comment:
If we exit from here, what happens to the directory we just created in this function? It seems that in case of error what you are doing is repetitive and might want all of these returns and close to go in a separate function. This might also help you clear the directory if not needed.